pennylane/grouping/group_observables.py

```diff
@@ -0,0 +1,227 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the high-level Pauli word partitioning functionality used in measurement optimization.
+"""
+
+from pennylane.grouping.utils import (
+    convert_observables_to_binary,
+    binary_to_pauli,
+    are_identical_pauli_words,
+    get_qwc_compliment_adj_matrix,
+)
+from pennylane.grouping.graph_colouring import largest_first, recursive_largest_first
+import numpy as np
+
+GROUPING_TYPES = ["qwc", "commuting", "anticommuting"]
+GRAPH_COLOURING_METHODS = {"lf": largest_first, "rlf": recursive_largest_first}
+
+
+class PauliGroupingStrategy:
+    """
+    Class for partitioning a list of Pauli words according to some binary symmetric relation.
+
+    Partitions are defined by the binary symmetric relation of interest, e.g. all Pauli words in a
+    partition being mutually commuting. The partition is accomplished by formulating the list of
+    observables as a graph where nodes represent individual Pauli words and connections between
+    nodes corresponds to the two Pauli words satisfying the symmetric binary relation. Obtaining
+    the fewest number of partitions to cover the list of Pauli words can then be accomplished by
+    graph coloring on the complimentary graph.
+
+    Args:
+        observables (list[Observable]): A list of Pauli words to be partitioned according to a
+        grouping strategy.
+
+    Keyword Args:
+        grouping_type (str): The binary relation used to define partitions of the Pauli words.
+        graph_colourer (str): The heuristic algorithm to employ for graph colouring.
+
+    Raises:
+        ValueError: if `grouping_type` or `graph_colourer` are not recognized as elements of
+        `GROUPING_TYPES` or `GRAPH_COLOURING_METHODS` respectively.
+
+    """
+
+    def __init__(self, observables, grouping_type="qwc", graph_colourer="rlf"):
+
+        if grouping_type.lower() not in GROUPING_TYPES:
+            raise ValueError(
+                "Grouping type must be one of: {}, instead got {}.".format(
+                    GROUPING_TYPES, grouping_type
+                )
+            )
+
+        self.grouping_type = grouping_type.lower()
+
+        if graph_colourer.lower() not in GRAPH_COLOURING_METHODS.keys():
+            raise ValueError(
+                "Graph colouring method must be one of: {}, instead got {}.".format(
+                    list(GRAPH_COLOURING_METHODS.keys()), graph_colourer
+                )
+            )
+
+        self.graph_colourer = GRAPH_COLOURING_METHODS[graph_colourer.lower()]
+        self.observables = observables
+        self.binary_observables = None
+        self.adj_matrix = None
+        self.grouped_paulis = None
+
+    def obtain_binary_repr(self, n_qubits=None, wire_map=None):
+        """Converts the list of Pauli words to a binary matrix.
+
```

Comment:
Can you add the arguments to the comments for this function?

Reply:
Thanks, also added the missing arguments description to the docstrings for `convert_observables_to_binary` function in `utils.py`.